ethcore/src/evm/wasm/runtime.rs

```diff
@@ -0,0 +1,356 @@
+		Ok(previous_top.into())
+	}
+
+	/// Report gas cost with the params passed in wasm stack
+	fn gas(&mut self, context: interpreter::CallerContext) 
+		-> Result<Option<interpreter::RuntimeValue>, interpreter::Error> 
+	{
+		let amount = context.value_stack.pop_as::<i32>()? as u64;
+		if self.charge_gas(amount) {
+			Ok(None)
+		} else {
+			Err(interpreter::Error::Trap(format!("Gas exceeds limits of {}", self.gas_limit)))			
+		}
+	}
+
+	fn charge_gas(&mut self, amount: u64) -> bool {
+		let prev = self.gas_counter;
+		if prev + amount > self.gas_limit {
+			// exceeds gas
+			false
+		} else {
+			self.gas_counter = prev + amount;
+			true
+		}		
+	}
+
+	fn h256_at(&self, ptr: WasmPtr) -> Result<H256, interpreter::Error> {
+		Ok(H256::from_slice(&ptr.slice(32, &*self.memory)
+			.map_err(|_| interpreter::Error::Trap("Memory access violation".to_owned()))?
+		))	
+	}
+
+	fn pop_h256(&self, context: &mut interpreter::CallerContext) -> Result<H256, interpreter::Error> {
+		let ptr = WasmPtr::from_i32(context.value_stack.pop_as::<i32>()?)
+			.map_err(|_| interpreter::Error::Trap("Memory access violation".to_owned()))?;
+		self.h256_at(ptr)
+	}
+
+	fn pop_u256(&self, context: &mut interpreter::CallerContext) -> Result<U256, interpreter::Error> {
+		let ptr = WasmPtr::from_i32(context.value_stack.pop_as::<i32>()?)
+			.map_err(|_| interpreter::Error::Trap("Memory access violation".to_owned()))?;
+		self.h256_at(ptr).map(Into::into)
+	}
+
+	fn address_at(&self, ptr: WasmPtr) -> Result<Address, interpreter::Error> {
+		Ok(Address::from_slice(&ptr.slice(20, &*self.memory)
+			.map_err(|_| interpreter::Error::Trap("Memory access violation".to_owned()))?
+		))	
+	}
+
+	fn pop_address(&self, context: &mut interpreter::CallerContext) -> Result<Address, interpreter::Error> {
+		let ptr = WasmPtr::from_i32(context.value_stack.pop_as::<i32>()?)
+			.map_err(|_| interpreter::Error::Trap("Memory access violation".to_owned()))?;
+		self.address_at(ptr)
+	}
+
+	fn user_trap(&mut self, _context: interpreter::CallerContext) 
+		-> Result<Option<interpreter::RuntimeValue>, interpreter::Error> 
+	{
+		Err(interpreter::Error::Trap("unknown trap".to_owned()))
+	}
+
+	fn user_noop(&mut self, 
+		_context: interpreter::CallerContext
+	) -> Result<Option<interpreter::RuntimeValue>, interpreter::Error> {
+		Ok(None)
+	}
+
+	/// Write call descriptor to wasm memory
+	pub fn write_descriptor(&mut self, call_args: CallArgs) -> Result<WasmPtr, Error> {
+		let d_ptr = self.alloc(16)?;
+
+		let args_len = call_args.len();
+		let args_ptr = self.alloc(args_len)?;
+
+		// write call descriptor
+		// call descriptor is [args_ptr, args_len, return_ptr, return_len]
+		//   all are 4 byte length, last 2 are zeroed
+		let mut d_buf = [0u8; 16];
+		LittleEndian::write_u32(&mut d_buf[0..4], args_ptr);
+		LittleEndian::write_u32(&mut d_buf[4..8], args_len);
+		self.memory.set(d_ptr, &d_buf)?;
+
+		// write call args to memory
+		self.memory.set(args_ptr, &call_args.address)?;
```

Comment:
Sounds like a great idea. Don't get me wrong, the code looks good, I'm just pointing to some areas that might deserve a second look. There's nothing that I would argue against merging because of.